lib/srv/db/mysql/engine.go

```diff
@@ -0,0 +1,223 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysql
+
+import (
+	"context"
+	"net"
+	"strings"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+	"github.com/gravitational/teleport/lib/srv/db/mysql/protocol"
+
+	"github.com/siddontang/go-mysql/client"
+	"github.com/siddontang/go-mysql/packet"
+	"github.com/siddontang/go-mysql/server"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+)
+
+// Engine implements the MySQL database service that accepts client
+// connections coming over reverse tunnel from the proxy and proxies
+// them between the proxy and the MySQL database instance.
+//
+// Implements common.Engine.
+type Engine struct {
+	// Auth handles database access authentication.
+	Auth *common.Auth
+	// Audit emits database access audit events.
+	Audit *common.Audit
+	// Context is the database server close context.
+	Context context.Context
+	// Clock is the clock interface.
+	Clock clockwork.Clock
+	// Log is used for logging.
+	Log logrus.FieldLogger
+}
+
+// HandleConnection processes the connection from MySQL proxy coming
+// over reverse tunnel.
+//
+// It handles all necessary startup actions, authorization and acts as a
+// middleman between the proxy and the database intercepting and interpreting
+// all messages i.e. doing protocol parsing.
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session, clientConn net.Conn) (err error) {
+	// Make server conn to get access to protocol's WriteOK/WriteError methods.
+	proxyConn := server.Conn{Conn: packet.NewConn(clientConn)}
+	defer func() {
+		if err != nil {
+			if err := proxyConn.WriteError(err); err != nil {
+				e.Log.WithError(err).Error("Failed to send error to client.")
+			}
+		}
+	}()
+	// Perform authorization checks.
+	err = e.checkAccess(sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Establish connection to the MySQL server.
+	serverConn, err := e.connect(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		err := serverConn.Close()
+		if err != nil {
+			e.Log.WithError(err).Error("Failed to close connection to MySQL server.")
+		}
+	}()
+	// Send back OK packet to indicate auth/connect success. At this point
+	// the original client should consider the connection phase completed.
+	err = proxyConn.WriteOK(nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = e.Audit.OnSessionStart(e.Context, *sessionCtx, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		err := e.Audit.OnSessionEnd(e.Context, *sessionCtx)
+		if err != nil {
+			e.Log.WithError(err).Error("Failed to emit audit event.")
+		}
+	}()
+	// Copy between the connections.
+	clientErrCh := make(chan error, 1)
+	serverErrCh := make(chan error, 1)
+	go e.receiveFromClient(clientConn, serverConn, clientErrCh, sessionCtx)
```

Comment:
I thought about it as well, but the context won't be of much use there because what these handlers do comes down to reading from the client/service connections which is blocking and there's no good way to cancel those reads with a context. So both will exit when at least one side of the connection is closed because the other one will be closed as well and reads will start getting i/o errors.